lasagne/layers/recurrent.py

```diff
@@ -450,7 +450,14 @@ def step_masked(input_n, mask_n, hid_previous, *args):
             return [hid_out]
 
         if mask is not None:
-            mask = mask.dimshuffle(1, 0, 'x')
+            # After processing, the input has shape
+            # (seq_len, batch_size, f_0, f_1, ..., f_N)
+            # We have to determine N, which is the number of feature dimensions
+            # in order to adjust the shape of the masking variable to
+            # (seq_len, batch_size, 1, ..., 1)
+            #                       |---N---|
+            num_feature_dims = input.ndim - 2
+            mask = mask.dimshuffle((1, 0) + ('x',) * num_feature_dims)
```

Comment:
Good catch! To make the implementation simpler, though, could you instead do:
```python
if mask is not None:
    # expand mask from (seq_len, batch_size) to cover all feature dimensions
    mask = T.shape_padright(mask, input.ndim - 2)
```
This is easier to read and also doesn't require a long comment to explain!

Reading again, your implementation is actually wrong -- the mask needs to match the dimensionality of the hidden states, not the input states. So instead of `input.ndim - 2`, use `len(self.input_to_hidden.output_shape) - 2`. Putting this together:
```python
if mask is not None:
    # expand mask from (seq_len, batch_size) to cover all hidden dimensions
    hid_ndim = len(self.input_to_hidden.output_shape)
    mask = T.shape_padright(mask, hid_ndim - 2)
```

Could you update your PR accordingly? This would be great!